### PR TITLE
[#1119] improvement: Explicitly throw BUFFER_LIMIT_OF_HUGE_PARTITION instead of NO_BUFFER

### PR DESCRIPTION
**Pull request by lifeSo**
### What changes were proposed in this pull request?

Explicitly throw BUFFER_LIMIT_OF_HUGE_PARTITION instead of NO_BUFFER

### Why are the changes needed?

Fix: # https://github.com/apache/incubator-uniffle/issues/1119

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

test on prd and unit test case

### CODE REVIEW

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java [hunk 1]
```diff
@@ -458,31 +459,37 @@ public long requireBuffer(
       String appId, int shuffleId, List<Integer> partitionIds, int requireSize) {
     ShuffleTaskInfo shuffleTaskInfo = shuffleTaskInfos.get(appId);
     if (null == shuffleTaskInfo) {
-      return RequireBufferStatusCode.NO_REGISTER.statusCode();
+      LOG.error("Find not registered app, appId: {}, shuffleId: {}", appId, shuffleId);
```
Comment (zuston) [inline server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:462]: How about `no such app is registered. appId: {}, shuffleId: {}`

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java [hunk 2]
```diff
@@ -458,31 +459,37 @@ public long requireBuffer(
...
+      throw new NoRegisterException("Not Registered, appId: " + appId);
     }
     for (int partitionId : partitionIds) {
       long partitionUsedDataSize = getPartitionDataSize(appId, shuffleId, partitionId);
       if (shuffleBufferManager.limitHugePartition(
           appId, shuffleId, partitionId, partitionUsedDataSize)) {
         ShuffleServerMetrics.counterTotalRequireBufferFailedForHugePartition.inc();
-        return RequireBufferStatusCode.NO_BUFFER.statusCode();
+        String errorMessage =
```
Comment (zuston) [inline server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:470]: Why not throwing the `NoBufferForHugePartitionException` ?

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java [hunk 3]
```diff
@@ -458,31 +459,37 @@ public long requireBuffer(
...
+            String.format(
+                "No Buffer For Huge Partition, appId: %s, shuffleId: %s, partitionIds: %s, partitionUsedDataSize: %s",
```
Comment (zuston) [inline server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:472]: How about `Huge partition is limited to writing. appId: %s, shuffleId: %s, partitionIds: %s, partitionUsedDataSize: %s` ?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java [hunk 4]
```diff
@@ -393,28 +394,30 @@ public void requireBuffer(
...
+                    appId,
+                    request.getShuffleId(),
+                    request.getPartitionIdsList(),
+                    request.getRequireSize());
+      }
+    } catch (NoRegisterException e) {
+      requireBufferId = -1;
       status = StatusCode.NO_REGISTER;
       ShuffleServerMetrics.counterTotalRequireBufferFailed.inc();
+    } catch (NoBufferException e) {
```
Comment (zuston) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java:417]: How to handle `NoBufferForHugePartitionException,` I don't see these code/.

**Review by zuston — COMMENTED** [review]
Could you add some tests for huge partition no buffer exception ?

**Comment by lifeSo** [comment]
> Could you add some tests for huge partition no buffer exception ?

Ok, I added sendDataAndRequireBufferTest() test case in ShuffleServerGrpcTest. Mainly:
image

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java [hunk 5]
```diff
@@ -393,28 +395,30 @@ public void requireBuffer(
...
+                    appId,
+                    request.getShuffleId(),
+                    request.getPartitionIdsList(),
+                    request.getRequireSize());
+      }
+    } catch (NoRegisterException e) {
+      requireBufferId = -1;
       status = StatusCode.NO_REGISTER;
       ShuffleServerMetrics.counterTotalRequireBufferFailed.inc();
+    } catch (NoBufferException | NoBufferForHugePartitionException e) {
```
Comment (zuston) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java:418]: This looks weird, why not introducing new stateCode of `NoBufferForHugePartitionException`

**Review by zuston — COMMENTED** [review]
I hope the memory message of being limited by huge partition should be shown in client side rather than only having one exception in server

File: common/src/main/java/org/apache/uniffle/common/rpc/StatusCode.java [hunk 6]
```diff
@@ -27,13 +27,14 @@ public enum StatusCode {
...
   DOUBLE_REGISTER(1),
   NO_BUFFER(2),
-  INVALID_STORAGE(3),
-  NO_REGISTER(4),
-  NO_PARTITION(5),
-  INTERNAL_ERROR(6),
-  TIMEOUT(7),
-  ACCESS_DENIED(8),
-  INVALID_REQUEST(9),
+  NO_BUFFER_FOR_HUGE_PARTITION(3),
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/rpc/StatusCode.java:30]: You should consider the compatility. 
```
NO_BUFFER_FOR_HUGE_PARTITION(10)
```

Comment (zuston) [reply common/src/main/java/org/apache/uniffle/common/rpc/StatusCode.java:30]: +1. should be 11

File: integration-test/common/src/test/java/org/apache/uniffle/test/IntegrationTestBase.java [hunk 7]
```diff
@@ -58,6 +58,8 @@ public abstract class IntegrationTestBase extends HadoopTestBase {
   protected static final int JETTY_PORT_1 = 19998;
   protected static final int JETTY_PORT_2 = 20040;
   protected static final String COORDINATOR_QUORUM = LOCALHOST + ":" + COORDINATOR_PORT_1;
+  protected static final String SHUFFLE_SERVER_METRICS_URL =
```
Comment (zuston) [inline integration-test/common/src/test/java/org/apache/uniffle/test/IntegrationTestBase.java:61]: why need this?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerMetrics.java [hunk 8]
```diff
@@ -80,7 +80,9 @@ public class ShuffleServerMetrics {
   private static final String LOCAL_DISK_PATH_LABEL = "local_disk_path";
   public static final String LOCAL_DISK_PATH_LABEL_ALL = "ALL";
   private static final String TOTAL_REQUIRE_BUFFER_FAILED = "total_require_buffer_failed";
-  private static final String TOTAL_REQUIRE_BUFFER_FAILED_FOR_HUGE_PARTITION =
+  private static final String TOTAL_REQUIRE_BUFFER_FAILED_FOR_NO_REGISTER =
```
Comment (zuston) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerMetrics.java:83]: this is not related with this PR. Right?

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java [hunk 9]
```diff
@@ -223,7 +223,8 @@ public long requirePreAllocation(
     long result = FAILED_REQUIRE_ID;
     Random random = new Random();
     final int backOffBase = 2000;
-    while (rpcResponse.getStatus() == RssProtos.StatusCode.NO_BUFFER) {
+    while (rpcResponse.getStatus() == RssProtos.StatusCode.NO_BUFFER
+        || rpcResponse.getStatus() == RssProtos.StatusCode.NO_BUFFER_FOR_HUGE_PARTITION) {
```
Comment (zuston) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:227]: Could you expose the concrete status code into log message?

**Comment by jerqi** [comment]
Could you reply all the comments from zuston?

**Comment by lifeSo** [comment]
> **[zuston](/zuston) ** reviewed [ Dec 28, 2023 ](#pullrequestreview-1797951840)
> [ View reviewed changes ](/apache/incubator-uniffle/pull/1393/files/f89872ca15ab704b0cbe3b5c23382f8ce6d1d16f)



> Could you reply all the comments from zuston?

@jerqi  I checked and find one place miss modify, and I corrected it. 
I thin all comments is reply and corrected.

**Comment by jerqi** [comment]
> > **[zuston](/zuston) ** reviewed [ Dec 28, 2023 ](#pullrequestreview-1797951840)
> > [ View reviewed changes ](/apache/incubator-uniffle/pull/1393/files/f89872ca15ab704b0cbe3b5c23382f8ce6d1d16f)
> 
> > Could you reply all the comments from zuston?
> 
> @jerqi I checked and find one place miss modify, and I corrected it. I thin all comments is reply and corrected.

You should reply the comment instead of marking the comment resolved. The reply is the your solution. Maybe it will be very simple like `fixed`. Let reviewer mark the comment resolved.

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java (hunk 9 above)
Comment (zuston) [reply internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:237]: ping @lifeSo

**Comment by zuston** [comment]
#1425 Merged. And close this.
